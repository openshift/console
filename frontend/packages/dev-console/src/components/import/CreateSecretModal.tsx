import * as React from 'react';
import {
  withSecretForm,
  ImageSecretForm,
  SourceSecretForm,
  SecretTypeAbstraction,
} from '@console/internal/components/secrets/create-secret';
import {
  createModalLauncher,
  ModalComponentProps,
} from '@console/internal/components/factory/modal';

export interface CreateSecretModalProps {
  save?: (name: string) => void;
  namespace: string;
  secretType: SecretTypeAbstraction;
}

type Props = CreateSecretModalProps & ModalComponentProps;

const getSecretForm = (type: SecretTypeAbstraction) => {
  switch (type) {
    case SecretTypeAbstraction.source:
      return withSecretForm(SourceSecretForm, true);
    case SecretTypeAbstraction.image:
      return withSecretForm(ImageSecretForm, true);
    default:
      return null;
  }
};

const CreateSecretModal: React.FC<Props> = ({ close, namespace, save, secretType }) => {
  const handleSave = (name: string) => {
    close();
    save(name);
  };

  const CreateSecretForm = getSecretForm(secretType);
  return (
    <CreateSecretForm
      onCancel={close}
      onSave={handleSave}
      fixed={{ metadata: { namespace } }}
      secretTypeAbstraction={secretType}
      titleVerb="Create"
      isCreate
    />
  );
};

export const secretModalLauncher = createModalLauncher<Props>(CreateSecretModal);

export default CreateSecretModal;
