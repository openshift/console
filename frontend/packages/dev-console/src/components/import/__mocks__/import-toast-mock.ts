import { DeploymentModel, RouteModel } from '@console/internal/models';
import { RouteKind, K8sResourceKind } from '@console/internal/module/k8s';

type DeployedResource = K8sResourceKind & {
  spec: {
    selector?: string;
    template: string;
  };
};

export const mockResources: { deployedResources: DeployedResource[]; route?: RouteKind }[] = [
  {
    deployedResources: [{ kind: DeploymentModel.kind, spec: { selector: null, template: null } }],
    route: { kind: RouteModel.kind, spec: { host: 'testURL', to: null } },
  },
  { deployedResources: [{ kind: DeploymentModel.kind, spec: { selector: null, template: null } }] },
];
