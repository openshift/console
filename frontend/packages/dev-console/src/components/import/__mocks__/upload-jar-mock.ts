import { ImportStrategy } from '@console/git-service/src';
import { UNASSIGNED_KEY } from '@console/topology/src/const';
import { healthChecksProbeInitialData } from '../../health-checks/health-checks-probe-utils';
import {
  InsecureTrafficType,
  Resources,
  TerminationType,
  UploadJarFormData,
} from '../import-types';
import { serverlessInitialValues } from './serverless-mock';

export const uploadJarMockFormData: UploadJarFormData = {
  project: {
    name: 'my-app',
    displayName: '',
    description: '',
  },
  application: {
    initial: '',
    name: '',
    selectedKey: UNASSIGNED_KEY,
  },
  name: 'java-ex-git',
  fileUpload: {
    name: 'springApp.jar',
    value: '',
  },
  import: {
    selectedStrategy: {
      name: '',
      type: ImportStrategy.S2I,
      priority: 0,
      detectedFiles: [],
    },
  },
  image: {
    selected: 'java',
    recommended: '',
    tag: 'openjdk-11-el7',
    tagObj: {
      name: 'openjdk-11-el7',
      annotations: {
        description: 'Build and run Java applications using Maven and OpenJDK 11.',
        iconClass: 'icon-rh-openjdk',
        'openshift.io/display-name': 'Red Hat OpenJDK 11 (RHEL 7)',
        sampleContextDir: 'undertow-servlet',
        sampleRepo: 'https://github.com/jboss-openshift/openshift-quickstarts',
        supports: 'java:11,java',
        tags: 'builder,java,openjdk',
        version: '11',
      },
      from: { kind: 'DockerImage', name: 'registry.redhat.io/openjdk/openjdk-11-rhel7:latest' },
      generation: 2,
      importPolicy: {},
      referencePolicy: { type: 'Local' },
    },
    ports: [
      { containerPort: 8080, protocol: 'TCP' },
      { containerPort: 8443, protocol: 'TCP' },
      { containerPort: 8778, protocol: 'TCP' },
    ],
    isRecommending: false,
    couldNotRecommend: false,
  },
  serverless: serverlessInitialValues,
  route: {
    create: true,
    targetPort: '',
    path: '',
    hostname: '',
    secure: false,
    tls: {
      termination: 'None' as TerminationType,
      insecureEdgeTerminationPolicy: 'None' as InsecureTrafficType,
      caCertificate: '',
      certificate: '',
      destinationCACertificate: '',
      key: '',
    },
  },
  resources: Resources.OpenShift,
  build: {
    env: [],
    triggers: {},
    strategy: 'Source',
  },
  deployment: {
    env: [],
    triggers: {
      image: true,
      config: true,
    },
    replicas: 1,
  },
  labels: {},
  limits: {
    cpu: {
      request: '',
      requestUnit: 'm',
      defaultRequestUnit: 'm',
      limit: '',
      limitUnit: 'm',
      defaultLimitUnit: 'm',
    },
    memory: {
      request: '',
      requestUnit: 'Mi',
      defaultRequestUnit: 'Mi',
      limit: '',
      limitUnit: 'Mi',
      defaultLimitUnit: 'Mi',
    },
  },
  healthChecks: healthChecksProbeInitialData,
};
