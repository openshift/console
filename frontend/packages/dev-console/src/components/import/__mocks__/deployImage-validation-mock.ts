import { DeployImageFormData, Resources } from '../import-types';

export const mockDeployImageFormData: DeployImageFormData = {
  project: {
    name: 'mock-project',
    displayName: '',
    description: '',
  },
  application: {
    initial: 'mock-app',
    name: 'mock-app',
    selectedKey: 'mock-app',
  },
  name: 'test-app',
  searchTerm: 'test-app',
  registry: 'external',
  imageStream: {
    image: '',
    tag: '',
    namespace: '',
  },
  isi: {
    image: {
      dockerImageLayers: [
        {
          name: 'sha256:8f91359f1fffbf32b24ca854fb263d88a222371f38e90cf4583c5742cfdc3039',
          size: 22510654,
          mediaType: 'application/vnd.docker.image.rootfs.diff.tar.gzip',
        },
        {
          name: 'sha256:e6e554c0af6fbe639e85531e90ab893e221666cb8c9c87e5e260a3cebadfc4da',
          size: 4501229,
          mediaType: 'application/vnd.docker.image.rootfs.diff.tar.gzip',
        },
      ],
      dockerImageManifestMediaType: 'application/vnd.docker.distribution.manifest.v2+json',
      dockerImageMetadata: {
        Architecture: 'amd64',
        Config: {
          cmd: ['test'],
          Entrypoint: ['docker-entrypoint.sh'],
          Image: 'sha256:231b50bbbc3f4606a0a7c527c63f9d447e7c18592433bf82ad4692787bf925ab',
        },
        Container: 'e03bd283544a70575c453759ead333056c2112e485ff51b2df4656d28e27c6dd',
        Created: '2019-09-12T04:24:43Z',
        DockerVersion: '18.06.1-ce',
        Id: 'sha256:b8fd9553f1f06e56173db706cf47dfc64ae3a7aeb213cc0e4fc476983b62bf16',
        Size: 129977572,
        apiVersion: '1.0',
        kind: 'DockerImage',
        dockerImageMetadataVersion: '1.0',
        dockerImageReference:
          'mysql@sha256:2e4114bdc9dd797549f6df0cffb5f6cb6354bef9d96223a5935b6b17aea03840',
      },
      metadata: {
        annotations: {
          'image.openshift.io/dockerLayersOrder': 'ascending',
        },
        creationTimestamp: null,
        name: 'sha256:2e4114bdc9dd797549f6df0cffb5f6cb6354bef9d96223a5935b6b17aea03840',
      },
    },
    name: 'myimage',
    tag: 'latest',
    status: { metadata: {}, status: 'success' },
    ports: [],
  },
  image: {
    name: '',
    image: {},
    tag: 'latest',
    status: { metadata: {}, status: '' },
    ports: [],
  },
  isSearchingForImage: false,
  resources: Resources.OpenShift,
  serverless: {
    scaling: {
      minpods: 0,
      maxpods: '',
      concurrencytarget: '',
      concurrencylimit: '',
    },
  },
  route: {
    create: true,
    targetPort: '',
    path: '',
    hostname: '',
    secure: false,
    tls: {
      termination: '',
      insecureEdgeTerminationPolicy: '',
      caCertificate: '',
      certificate: '',
      destinationCACertificate: '',
      privateKey: '',
    },
  },
  build: {
    env: [],
    triggers: {
      webhook: true,
      image: true,
      config: true,
    },
    strategy: 'Source',
  },
  deployment: {
    env: [],
    triggers: {
      image: true,
      config: true,
    },
    replicas: 1,
  },
  labels: {},
  env: {},
  limits: {
    cpu: {
      request: null,
      requestUnit: 'm',
      limit: null,
      limitUnit: 'm',
    },
    memory: {
      request: null,
      requestUnit: 'Mi',
      limit: null,
      limitUnit: 'Mi',
    },
  },
};
