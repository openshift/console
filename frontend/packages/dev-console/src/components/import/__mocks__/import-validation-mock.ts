import { ValidatedOptions } from '@patternfly/react-core';
import { GitImportFormData, Resources } from '../import-types';

export const mockFormData: GitImportFormData = {
  name: 'test-app',
  project: {
    name: 'mock-project',
    displayName: '',
    description: '',
  },
  application: {
    initial: 'mock-app',
    name: 'mock-app',
    selectedKey: 'mock-app',
  },
  git: {
    url: 'https://github.com/test/repo',
    type: 'github',
    ref: '',
    dir: '',
    showGitType: false,
    secret: '',
    urlValidation: ValidatedOptions.default,
    isUrlValidating: false,
  },
  docker: {
    dockerfilePath: 'Dockerfile',
    containerPort: 8080,
  },
  image: {
    selected: 'nodejs',
    recommended: '',
    tag: 'latest',
    tagObj: {},
    ports: [],
    isRecommending: false,
    couldNotRecommend: false,
  },
  route: {
    create: false,
    targetPort: '',
    path: '',
    hostname: '',
    secure: false,
    tls: {
      termination: '',
      insecureEdgeTerminationPolicy: '',
      caCertificate: '',
      certificate: '',
      destinationCACertificate: '',
      privateKey: '',
    },
  },
  build: {
    env: [],
    triggers: {
      webhook: true,
      image: true,
      config: true,
    },
    strategy: 'Source',
  },
  deployment: {
    env: [],
    triggers: {
      image: true,
      config: true,
    },
    replicas: 1,
  },
  labels: {},
  resources: Resources.OpenShift,
  serverless: {
    scaling: {
      minpods: 0,
      maxpods: '',
      concurrencytarget: '',
      concurrencylimit: '',
    },
  },
  limits: {
    cpu: {
      request: null,
      requestUnit: 'm',
      defaultRequestUnit: 'm',
      limit: null,
      limitUnit: 'm',
      defaultLimitUnit: 'm',
    },
    memory: {
      request: null,
      requestUnit: 'Mi',
      defaultRequestUnit: 'Mi',
      limit: null,
      limitUnit: 'Mi',
      defaultLimitUnit: 'Mi',
    },
  },
};
