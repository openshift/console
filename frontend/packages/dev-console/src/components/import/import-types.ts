import { ValidatedOptions } from '@patternfly/react-core';
import { K8sResourceKind, ContainerPort } from '@console/internal/module/k8s';
import { LazyLoader } from '@console/plugin-sdk';
import { NameValuePair, NameValueFromPair } from '@console/shared';
import { NormalizedBuilderImages } from '../../utils/imagestream-utils';

export interface DeployImageFormProps {
  builderImages?: NormalizedBuilderImages;
  projects?: FirehoseList;
}
export type ImageStreamPayload = boolean | K8sResourceKind;

export type ImageStreamState = {
  hasAccessToPullImage: ImageStreamPayload;
  accessLoading: ImageStreamPayload;
  loading: ImageStreamPayload;
  hasCreateAccess: ImageStreamPayload;
  selectedImageStream: ImageStreamPayload;
};
export enum ImageStreamActions {
  setAccessLoading = 'setAccessLoading',
  setLoading = 'setLoading',
  setSelectedImageStream = 'setSelectedImageStream',
  setHasAccessToPullImage = 'setHasAccessToPullImage',
  setHasCreateAccess = 'setHasCreateAccess',
}
export type ImageStreamAction = { type: ImageStreamActions; value: ImageStreamPayload };
export interface ImageStreamContextProps {
  state: ImageStreamState;
  dispatch: React.Dispatch<ImageStreamAction>;
  hasImageStreams: boolean;
  setHasImageStreams: (value: boolean) => void;
  setValidated: (validated: ValidatedOptions) => void;
}
export interface SourceToImageFormProps {
  builderImages?: NormalizedBuilderImages;
  projects?: {
    data: [];
    loaded: boolean;
  };
}

export interface GitImportFormProps {
  builderImages?: NormalizedBuilderImages;
  projects?: {
    data: [];
    loaded: boolean;
  };
}

export interface FirehoseList {
  data?: K8sResourceKind[];
  [key: string]: any;
}

export interface DeployImageFormData {
  formType?: string;
  project: ProjectData;
  application: ApplicationData;
  name: string;
  searchTerm: string;
  registry: string;
  imageStream: {
    image: string;
    tag: string;
    namespace: string;
    grantAccess?: boolean;
  };
  isi: ImageStreamImageData;
  image: ImageStreamImageData;
  isSearchingForImage: boolean;
  resources: Resources;
  serverless?: ServerlessData;
  pipeline?: PipelineData;
  labels: { [name: string]: string };
  env?: { [name: string]: string };
  route: RouteData;
  build: BuildData;
  deployment: DeploymentData;
  limits: LimitsData;
}

export interface GitImportFormData {
  formType?: string;
  name: string;
  project: ProjectData;
  application: ApplicationData;
  git: GitData;
  docker: DockerData;
  serverless?: ServerlessData;
  pipeline?: PipelineData;
  image: ImageData;
  route: RouteData;
  resources: Resources;
  build: BuildData;
  deployment: DeploymentData;
  labels: { [name: string]: string };
  limits: LimitsData;
}

export interface ApplicationData {
  initial?: string;
  name: string;
  selectedKey: string;
}

export interface ImageData {
  selected: string;
  recommended: string;
  isRecommending: boolean;
  couldNotRecommend: boolean;
  tag: string;
  tagObj: object;
  ports: ContainerPort[];
}

export interface ImageStreamImageData {
  name: string;
  image: { [key: string]: any };
  tag: string;
  status: { metadata: {}; status: string };
  ports: ContainerPort[];
}

export interface ProjectData {
  name: string;
  displayName?: string;
  description?: string;
}

export interface GitData {
  url: string;
  type: string;
  ref: string;
  dir: string;
  showGitType: boolean;
  secret: string;
  urlValidation: ValidatedOptions;
  isUrlValidating: boolean;
}

export interface DockerData {
  dockerfilePath?: string;
  containerPort?: number;
}

export interface RouteData {
  disable?: boolean;
  create: boolean;
  targetPort: string;
  unknownTargetPort?: string;
  defaultUnknownPort?: number;
  path: string;
  hostname: string;
  secure: boolean;
  tls: TLSData;
}

export interface TLSData {
  termination: string;
  insecureEdgeTerminationPolicy: string;
  certificate: string;
  privateKey: string;
  caCertificate: string;
  destinationCACertificate: string;
}

export interface BuildData {
  triggers: {
    webhook?: boolean;
    image?: boolean;
    config?: boolean;
  };
  env: (NameValuePair | NameValueFromPair)[];
  strategy: string;
}

export interface DeploymentData {
  triggers: {
    image: boolean;
    config: boolean;
  };
  replicas: number;
  env: (NameValuePair | NameValueFromPair)[];
}

export interface ServerlessData {
  scaling: ServerlessScaling;
}

export interface PipelineData {
  enabled: boolean;
  template?: K8sResourceKind;
}

export interface ServerlessScaling {
  minpods: number;
  maxpods: number | '';
  concurrencytarget: number | '';
  concurrencylimit: number | '';
}

export enum GitTypes {
  github = 'github',
  gitlab = 'gitlab',
  bitbucket = 'bitbucket',
  unsure = 'other',
  invalid = '',
}

export const GitReadableTypes = {
  [GitTypes.github]: 'GitHub',
  [GitTypes.gitlab]: 'GitLab',
  [GitTypes.bitbucket]: 'Bitbucket',
  [GitTypes.unsure]: 'Other',
};

export enum ImportTypes {
  git = 'git',
  docker = 'docker',
  s2i = 's2i',
}

export enum Resources {
  OpenShift = 'openshift',
  Kubernetes = 'kubernetes',
  KnativeService = 'knative',
}

export interface ImportData {
  type: ImportTypes;
  title: string;
  buildStrategy: string;
  loader: LazyLoader<GitImportFormProps | SourceToImageFormProps>;
}

export enum TerminationTypes {
  edge = 'Edge',
  passthrough = 'Passthrough',
  reencrypt = 'Re-encrypt',
}

export enum InsecureTrafficTypes {
  None = 'None',
  Allow = 'Allow',
  Redirect = 'Redirect',
}

export enum PassthroughInsecureTrafficTypes {
  None = 'None',
  Redirect = 'Redirect',
}

export interface LimitsData {
  cpu: ResourceType;
  memory: ResourceType;
}

export interface ResourceType {
  request: number | string;
  requestUnit: string;
  defaultRequestUnit: string;
  limit: number | string;
  limitUnit: string;
  defaultLimitUnit: string;
}

export enum CPUUnits {
  m = 'millicores',
  '' = 'cores',
}

export enum MemoryUnits {
  Mi = 'Mi',
  Gi = 'Gi',
}

export enum ImportOptions {
  GIT = 'GIT',
  CONTAINER = 'CONTAINER',
  CATALOG = 'CATALOG',
  DOCKERFILE = 'DOCKERFILE',
  DATABASE = 'DATABASE',
}
