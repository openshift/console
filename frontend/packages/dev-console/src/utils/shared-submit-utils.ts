import * as _ from 'lodash';
import { K8sResourceKind } from '@console/internal/module/k8s';
import { GitImportFormData, DeployImageFormData } from '../components/import/import-types';
import { getAppLabels, getPodLabels, getAppAnnotations, mergeData } from './resource-label-utils';
import { makePortName } from './imagestream-utils';

export const annotations = {
  'openshift.io/generated-by': 'OpenShiftWebConsole',
};

export const dryRunOpt = { queryParams: { dryRun: 'All' } };

const isDeployImageFormData = (
  formData: DeployImageFormData | GitImportFormData,
): formData is DeployImageFormData => {
  return 'isi' in (formData as DeployImageFormData);
};

export const createService = (
  formData: DeployImageFormData | GitImportFormData,
  imageStreamData?: K8sResourceKind,
  originalService?: K8sResourceKind,
): K8sResourceKind => {
  const {
    project: { name: namespace },
    application: { name: application },
    name,
    labels: userLabels,
    image: { ports: imagePorts, tag: imageTag },
  } = formData;

  const imageStreamName = _.get(imageStreamData, 'metadata.name') || _.get(formData, 'image.name');
  const git = _.get(formData, 'git');

  const defaultLabels = getAppLabels(name, application, imageStreamName, imageTag);
  const podLabels = getPodLabels(name);
  const defaultAnnotations = git ? getAppAnnotations(git.url, git.ref) : annotations;

  let ports = imagePorts;
  if (_.get(formData, 'build.strategy') === 'Docker') {
    const port = { containerPort: _.get(formData, 'docker.containerPort'), protocol: 'TCP' };
    ports = [port];
  } else if (isDeployImageFormData(formData)) {
    const {
      isi: { ports: isiPorts },
    } = formData;
    ports = isiPorts;
  }

  const newService: any = {
    kind: 'Service',
    apiVersion: 'v1',
    metadata: {
      name,
      namespace,
      labels: { ...defaultLabels, ...userLabels },
      annotations: { ...defaultAnnotations },
    },
    spec: {
      selector: podLabels,
      ports: _.map(ports, (port) => ({
        port: port.containerPort,
        targetPort: port.containerPort,
        protocol: port.protocol,
        // Use the same naming convention as CLI new-app.
        name: makePortName(port),
      })),
    },
  };

  const service = mergeData(originalService, newService);

  return service;
};

export const createRoute = (
  formData: GitImportFormData | DeployImageFormData,
  imageStreamData?: K8sResourceKind,
  originalRoute?: K8sResourceKind,
): K8sResourceKind => {
  const {
    project: { name: namespace },
    application: { name: application },
    name,
    labels: userLabels,
    route: { hostname, secure, path, tls, targetPort: routeTargetPort },
    image: { ports: imagePorts, tag: imageTag },
  } = formData;

  const imageStreamName = _.get(imageStreamData, 'metadata.name') || _.get(formData, 'image.name');
  const git = _.get(formData, 'git');

  const defaultLabels = getAppLabels(name, application, imageStreamName, imageTag);
  const defaultAnnotations = git ? getAppAnnotations(git.url, git.ref) : annotations;

  let ports = imagePorts;
  if (isDeployImageFormData(formData)) {
    const {
      isi: { ports: isiPorts },
    } = formData;
    ports = isiPorts;
  }

  let targetPort;
  if (_.get(formData, 'build.strategy') === 'Docker') {
    const port = _.get(formData, 'docker.containerPort');
    targetPort = makePortName({
      containerPort: _.toInteger(port),
      protocol: 'TCP',
    });
  } else {
    targetPort = routeTargetPort || makePortName(_.head(ports));
  }

  const newRoute: any = {
    kind: 'Route',
    apiVersion: 'route.openshift.io/v1',
    metadata: {
      name,
      namespace,
      labels: { ...defaultLabels, ...userLabels },
      defaultAnnotations,
    },
    spec: {
      to: {
        kind: 'Service',
        name,
      },
      ...(secure ? { tls } : {}),
      host: hostname,
      path,
      // The service created by `createService` uses the same port as the container port.
      port: {
        // Use the port name, not the number for targetPort. The router looks
        // at endpoints, not services, when resolving ports, so port numbers
        // will not resolve correctly if the service port and container port
        // numbers don't match.
        targetPort,
      },
      wildcardPolicy: 'None',
    },
  };

  const route = mergeData(originalRoute, newRoute);

  return route;
};
