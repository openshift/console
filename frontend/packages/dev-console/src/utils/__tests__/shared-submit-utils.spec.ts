import * as _ from 'lodash';
import { mockDeployImageFormData } from '../../components/import/__mocks__/deployImage-validation-mock';
import { mockFormData } from '../../components/import/__mocks__/import-validation-mock';
import { GitImportFormData, DeployImageFormData } from '../../components/import/import-types';
import { createService, createRoute } from '../shared-submit-utils';

describe('Shared submit utils', () => {
  describe('Create Service', () => {
    it('should set the correct ports in service object', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);
      const mockDeployImageData: DeployImageFormData = _.cloneDeep(mockDeployImageFormData);
      const PORT = { containerPort: 8081, protocol: 'TCP' };
      let serviceObj;

      mockData.build.strategy = 'Docker';
      mockData.route.unknownTargetPort = '8080';
      serviceObj = createService(mockData);
      expect(serviceObj.spec.ports[0].port).toEqual(8080);

      mockDeployImageData.isi.ports = [PORT];
      serviceObj = createService(mockDeployImageData);
      expect(serviceObj.spec.ports[0].port).toEqual(8081);
    });

    it('should match the previous snapshot created with git import data', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);
      const serviceObj = createService(mockData);
      expect(serviceObj).toMatchSnapshot();
    });

    it('should match the previous snapshot created with deploy image data', () => {
      const mockDeployImageData: DeployImageFormData = _.cloneDeep(mockDeployImageFormData);
      const serviceObj = createService(mockDeployImageData);
      expect(serviceObj).toMatchSnapshot();
    });

    it('should expose only the custom port as TargetPort when it is set', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);

      mockData.build.strategy = 'Source';
      mockData.git.url = 'https://github.com/nodeshift-blog-examples/react-web-app';
      mockData.route.unknownTargetPort = '3000';
      const serviceObj = createService(mockData);
      const ports = serviceObj?.spec?.ports;

      expect(ports).not.toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            port: 8080,
          }),
        ]),
      );
    });

    it('While editing, the new custom route port must replace the current route port from services', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);

      mockData.build.strategy = 'Source';
      mockData.git.url = 'https://github.com/nodeshift-blog-examples/react-web-app';
      mockData.image.ports = [
        { containerPort: 8080, protocol: 'TCP' },
        { containerPort: 8081, protocol: 'TCP' },
        { containerPort: 8082, protocol: 'TCP' },
      ];
      mockData.route.unknownTargetPort = '8080';
      let serviceObj = createService(mockData);

      mockData.route.unknownTargetPort = '3000';
      serviceObj = createService(mockData);

      const ports = serviceObj?.spec?.ports;

      expect(ports).not.toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            port: 8080,
          }),
        ]),
      );
    });
  });

  describe('Create Route', () => {
    it('should set correct target port in route object', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);
      const mockDeployImageData: DeployImageFormData = _.cloneDeep(mockDeployImageFormData);
      let routeObj;

      mockData.build.strategy = 'Docker';
      mockData.route.unknownTargetPort = '8080';
      routeObj = createRoute(mockData);
      expect(routeObj.spec.port.targetPort).toEqual('8080-tcp');

      const PORT = { containerPort: 8082, protocol: 'TCP' };
      mockDeployImageData.isi.ports = [PORT];
      routeObj = createRoute(mockDeployImageData);
      expect(routeObj.spec.port.targetPort).toEqual('8082-tcp');

      mockDeployImageData.route.unknownTargetPort = '8081';
      routeObj = createRoute(mockDeployImageData);
      expect(routeObj.spec.port.targetPort).toEqual('8081-tcp');
    });

    it('should set correct route labels if custom labels are set', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);
      mockData.route.labels = {
        'custom-route-label': 'a-custom-route-label',
      };
      mockData.labels = {
        'shared-label': 'a-shared-label-value',
      };
      const routeObj = createRoute(mockData);
      expect(routeObj?.metadata?.labels).toEqual({
        app: 'test-app',
        'app.kubernetes.io/component': 'test-app',
        'app.kubernetes.io/instance': 'test-app',
        'app.kubernetes.io/name': 'test-app',
        'app.kubernetes.io/part-of': 'mock-app',
        'app.openshift.io/runtime-version': 'latest',
        'shared-label': 'a-shared-label-value',
        'custom-route-label': 'a-custom-route-label',
      });
    });

    it('should match the previous snapshot with git import data', () => {
      const mockData: GitImportFormData = _.cloneDeep(mockFormData);
      mockData.route.targetPort = '8080-tcp';
      const routeObj = createRoute(mockData);
      expect(routeObj).toMatchSnapshot();
    });

    it('should match the previous snapshot deploy image data', () => {
      const mockDeployImageData: DeployImageFormData = _.cloneDeep(mockDeployImageFormData);
      mockDeployImageData.route.targetPort = '8080-tcp';
      const routeObj = createRoute(mockDeployImageData);
      expect(routeObj).toMatchSnapshot();
    });
  });
});
