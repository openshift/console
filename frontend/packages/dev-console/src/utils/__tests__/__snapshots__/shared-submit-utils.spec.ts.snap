// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared submit utils Create Route should match the previous snapshot deploy image data 1`] = `
Object {
  "apiVersion": "route.openshift.io/v1",
  "kind": "Route",
  "metadata": Object {
    "defaultAnnotations": Object {
      "openshift.io/generated-by": "OpenShiftWebConsole",
    },
    "labels": Object {
      "app": "test-app",
      "app.kubernetes.io/component": "test-app",
      "app.kubernetes.io/instance": "test-app",
      "app.kubernetes.io/name": "",
      "app.kubernetes.io/part-of": "mock-app",
      "app.openshift.io/runtime": "",
      "app.openshift.io/runtime-version": "latest",
    },
    "name": "test-app",
    "namespace": "mock-project",
  },
  "spec": Object {
    "host": "",
    "path": "",
    "port": Object {
      "targetPort": "8080-tcp",
    },
    "to": Object {
      "kind": "Service",
      "name": "test-app",
    },
    "wildcardPolicy": "None",
  },
}
`;

exports[`Shared submit utils Create Route should match the previous snapshot with git import data 1`] = `
Object {
  "apiVersion": "route.openshift.io/v1",
  "kind": "Route",
  "metadata": Object {
    "defaultAnnotations": Object {
      "app.openshift.io/vcs-ref": "master",
      "app.openshift.io/vcs-uri": "https://github.com/test/repo",
      "openshift.io/generated-by": "OpenShiftWebConsole",
    },
    "labels": Object {
      "app": "test-app",
      "app.kubernetes.io/component": "test-app",
      "app.kubernetes.io/instance": "test-app",
      "app.kubernetes.io/name": undefined,
      "app.kubernetes.io/part-of": "mock-app",
      "app.openshift.io/runtime": undefined,
      "app.openshift.io/runtime-version": "latest",
    },
    "name": "test-app",
    "namespace": "mock-project",
  },
  "spec": Object {
    "host": "",
    "path": "",
    "port": Object {
      "targetPort": "8081-tcp",
    },
    "to": Object {
      "kind": "Service",
      "name": "test-app",
    },
    "wildcardPolicy": "None",
  },
}
`;

exports[`Shared submit utils Create Service should match the previous snapshot created with deploy image data 1`] = `
Object {
  "apiVersion": "v1",
  "kind": "Service",
  "metadata": Object {
    "annotations": Object {
      "openshift.io/generated-by": "OpenShiftWebConsole",
    },
    "labels": Object {
      "app": "test-app",
      "app.kubernetes.io/component": "test-app",
      "app.kubernetes.io/instance": "test-app",
      "app.kubernetes.io/name": "",
      "app.kubernetes.io/part-of": "mock-app",
      "app.openshift.io/runtime": "",
      "app.openshift.io/runtime-version": "latest",
    },
    "name": "test-app",
    "namespace": "mock-project",
  },
  "spec": Object {
    "ports": Array [],
    "selector": Object {
      "app": "test-app",
      "deploymentconfig": "test-app",
    },
  },
}
`;

exports[`Shared submit utils Create Service should match the previous snapshot created with git import data 1`] = `
Object {
  "apiVersion": "v1",
  "kind": "Service",
  "metadata": Object {
    "annotations": Object {
      "app.openshift.io/vcs-ref": "master",
      "app.openshift.io/vcs-uri": "https://github.com/test/repo",
      "openshift.io/generated-by": "OpenShiftWebConsole",
    },
    "labels": Object {
      "app": "test-app",
      "app.kubernetes.io/component": "test-app",
      "app.kubernetes.io/instance": "test-app",
      "app.kubernetes.io/name": undefined,
      "app.kubernetes.io/part-of": "mock-app",
      "app.openshift.io/runtime": undefined,
      "app.openshift.io/runtime-version": "latest",
    },
    "name": "test-app",
    "namespace": "mock-project",
  },
  "spec": Object {
    "ports": Array [],
    "selector": Object {
      "app": "test-app",
      "deploymentconfig": "test-app",
    },
  },
}
`;
